### PR TITLE
framework/task_manager: Add broadcast functionality

### PR DESCRIPTION
**Pull request by tyha1115**
Add broadcast functionality to task manager
The broadcast function can be used for broadcasting the specific pre-defined messages by using the task_manager_broadcast().
Pre-defined messages for broadcast are defined at the framework/include/task_manager/task_manager.h
In order to receive a broadcast message, user must register a handler to handle the broadcast message using task_manager_set_broadcast_handler() in advance.
When you set a handler for receiving broadcast messages, you can set a msg_mask to select message to receive (e.g., msg_mask = (WIFI_ON | WIFI_OFF)).

### CODE REVIEW

File: framework/src/task_manager/task_manager_broadcast.c [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <string.h>
```
Comment (sunghan-chang) [inline framework/src/task_manager/task_manager_broadcast.c]: First three headers looks unnecessary. Could you confirm?

File: framework/src/task_manager/task_manager_core.c [hunk 2]
```diff
@@ -447,6 +447,67 @@ static void taskmgr_termination_callback(void)
...
+
+static int taskmgr_get_handle_by_pid(int pid)
+{
+	int chk_idx;
+	for (chk_idx = 0; chk_idx < CONFIG_TASK_MANAGER_MAX_TASKS; chk_idx++) {
+		if (TASK_PID(chk_idx) == pid) {
+			return chk_idx;
+		}
+	}
+	return ERROR;
```
Comment (sunghan-chang) [inline framework/src/task_manager/task_manager_core.c]: Is ```ERROR``` enough? Can not this specify ```UNREGISTERED```?

File: framework/src/task_manager/task_manager_core.c [hunk 3]
```diff
@@ -447,6 +447,67 @@ static void taskmgr_termination_callback(void)
...
+		}
+	}
+}
+
+static int taskmgr_set_broadcast_handler(int msg_mask, int pid)
+{
+	int ret;
+	ret = taskmgr_get_handle_by_pid(pid);
+	if (ret < 0) {
+		return TM_FAIL_REGISTER;
```
Comment (sunghan-chang) [inline framework/src/task_manager/task_manager_core.c]: Is this ```FAIL_REGISTER```? Does this failure number intend originally?

File: framework/src/task_manager/task_manager_core.c [hunk 4]
```diff
@@ -447,6 +447,67 @@ static void taskmgr_termination_callback(void)
...
+	}
+	if (TASK_LIST_ADDR(ret) != NULL) {
+		TASK_MSG_MASK(ret) = msg_mask;
+		return OK;
+	}
+	return TM_FAIL_UNREGISTERED_TASK;
```
Comment (sunghan-chang) [inline framework/src/task_manager/task_manager_core.c]: When taskmgr_get_handle_by_pid returns fail, I think it is unregistered. Is there any chance to fail getting task address even though taskmgr_get_handle_by_pid returns OK?

File: framework/src/task_manager/task_manager_core.c [hunk 5]
```diff
@@ -447,6 +447,67 @@ static void taskmgr_termination_callback(void)
...
+}
+
+static void taskmgr_broadcast(int msg)
+{
+	int chk_idx;
+	int fd;
+	int ret;
+
+	fd = taskmgr_get_drvfd();
+	union sigval msg_broad;
```
Comment (sunghan-chang) [inline framework/src/task_manager/task_manager_core.c]: Let's define a local variable at start of function.

File: framework/src/task_manager/task_manager_core.c [hunk 6]
```diff
@@ -447,6 +447,67 @@ static void taskmgr_termination_callback(void)
...
+		if (TASK_PID(chk_idx) == pid) {
+			return chk_idx;
+		}
+	}
+	return ERROR;
+}
+
+static void taskmgr_broadcast(int msg)
+{
+	int chk_idx;
```
Comment (sunghan-chang) [inline framework/src/task_manager/task_manager_core.c]: chk_idx is not appropriate. task_idx is better I think because in your usages it means task..

File: framework/src/task_manager/task_manager_internal.h [hunk 7]
```diff
@@ -23,20 +23,21 @@
...
-#define TASKMGT_SCAN_GROUP           12
-#define TASKMGT_UNREGISTER_TASK      13
+#define TASKMGT_REGISTER_TASK           0
+#define TASKMGT_REGISTER_THREAD         1
+#define TASKMGT_START                   2
+#define TASKMGT_TERMINATE               3
+#define TASKMGT_RESTART                 4
+#define TASKMGT_PAUSE                   5
+#define TASKMGT_RESUME                  6
+#define TASKMGT_SCAN                    7
```
Comment (sunghan-chang) [inline framework/src/task_manager/task_manager_internal.h]: Is this used?

File: framework/src/task_manager/task_manager_set_broadcast_handler.c [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <stdio.h>
```
Comment (sunghan-chang) [inline framework/src/task_manager/task_manager_set_broadcast_handler.c]: You need to check header here as well.
Some headers are unncessary and some headers are missed.
For example, <unistd.h> for getpid().

File: framework/src/task_manager/task_manager_broadcast.c (hunk 1 above)
Comment (tyha1115) [reply framework/src/task_manager/task_manager_broadcast.c]: <string.h> is needed to use memset(), and others were deleted.

**Comment by seinfra** [comment]
Target : [3485a30d80da9ad9e4725f965da1dfa3457d778e] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [3485a30d80da9ad9e4725f965da1dfa3457d778e] - Code Rule Check Result:
framework/include/task_manager/task_manager.h:222: ERROR: [IDT_M_TAB] please, no spaces at the start of a line

**Comment by sunghan-chang** [comment]
@tyha1115 Please check a violation of coding rule.

File: framework/include/task_manager/task_manager.h [hunk 9]
```diff
@@ -200,7 +219,19 @@ int task_manager_unicast(int handle, void *msg, int msg_size, int timeout);
  * @return On success, OK is returned. On failure, defined negative value is returned.
  * @since TizenRT v2.0 PRE
  */
-int task_manager_set_handler(void (*func)(int signo, tm_msg_t *data));
+ int task_manager_set_handler(void (*func)(int signo, tm_msg_t *data));
```
Comment (sunghan-chang) [inline framework/include/task_manager/task_manager.h]: You added a space at start of line. Revert it.

File: framework/src/task_manager/task_manager_core.c [hunk 10]
```diff
@@ -447,6 +447,64 @@ static void taskmgr_termination_callback(void)
...
+{
+	if (msg_mask & TASK_MSG_MASK(handle)) {
+		return OK;
+	}
+	return ERROR;
+}
+
+static int taskmgr_get_handle_by_pid(int pid)
+{
+	int chk_idx;
```
Comment (sunghan-chang) [inline framework/src/task_manager/task_manager_core.c]: same as previous. task_idx or handle.

File: framework/src/task_manager/task_manager_core.c [hunk 11]
```diff
@@ -447,6 +447,64 @@ static void taskmgr_termination_callback(void)
...
+		}
+	}
+}
+
+static int taskmgr_set_broadcast_handler(int msg_mask, int pid)
+{
+	int ret;
+	ret = taskmgr_get_handle_by_pid(pid);
+	if (ret < 0) {
+		return TM_FAIL_UNREGISTERED_TASK;
```
Comment (sunghan-chang) [inline framework/src/task_manager/task_manager_core.c]: return ret;

File: framework/src/task_manager/task_manager_set_broadcast_handler.c [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <string.h>
+#include <signal.h>
+#include <stdlib.h>
```
Comment (sunghan-chang) [inline framework/src/task_manager/task_manager_set_broadcast_handler.c]: Why is this needed?

**Comment by seinfra** [comment]
Target : [da7a80681164388e51d3c1a670b9114502a6b9fa] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [da7a80681164388e51d3c1a670b9114502a6b9fa] - Code Rule Check (C++) OK.

File: framework/src/task_manager/task_manager_core.c [hunk 13]
```diff
@@ -447,6 +447,64 @@ static void taskmgr_termination_callback(void)
...
+}
+
+static void taskmgr_broadcast(int msg)
+{
+	int handle;
+	int fd;
+	int ret;
+	union sigval msg_broad;
+
+	fd = taskmgr_get_drvfd();
```
Comment (jeongchanKim) [inline framework/src/task_manager/task_manager_core.c]: please add the checking logic for fd

Comment (tyha1115) [reply framework/src/task_manager/task_manager_core.c]: Done.

**Comment by seinfra** [comment]
Target : [37e4aa015b88d583cd48d23de95fc4b1400d6073] -

**Comment by seinfra** [comment]
Target : [37e4aa015b88d583cd48d23de95fc4b1400d6073] -

**Comment by seinfra** [comment]
Target : [ced954bf87b9d82ac781fe28bb8246a30fe0165d] -

**Comment by seinfra** [comment]
Target : [ced954bf87b9d82ac781fe28bb8246a30fe0165d] -

**Comment by seinfra** [comment]
Target : [393533592d31d2ad9a91900697fd0e6c071af55f] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [393533592d31d2ad9a91900697fd0e6c071af55f] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [f601a8b0a1f0067f6f8b7ed8fbc806cbac882b7d] -

**Comment by seinfra** [comment]
Target : [f601a8b0a1f0067f6f8b7ed8fbc806cbac882b7d] -

**Comment by seinfra** [comment]
Target : [cd370c042702bb6c29f91142d0db68ca887aa95a] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [cd370c042702bb6c29f91142d0db68ca887aa95a] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [2cf045c1b9132be48e07a24a2c1f37af1ef6e618] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [2cf045c1b9132be48e07a24a2c1f37af1ef6e618] - Code Rule Check (C++) OK.
